### ISSUE TITLE
Unable to set margin on CollisionPolygon resulting in noticeable gaps when using Bullet engine

### ISSUE THREAD

**Issue by RobTheFiveNine** [godotengine/godot#45718]
**Godot version:**
v3.2.3.stable.official

**OS/device including version:**
Ubuntu 18.04

**Issue description:**
When a `RigidBody` with a `CollisionShape` collides with a `StaticBody` that contains a `CollisionPolygon`, a noticeable gap can be seen which seems to be the default margin that is applied. Unlike other collision objects, it's not possible to set the margin of the `CollisionPolygon` though.

A potential work around was mentioned in [This Similar Issue](https://github.com/godotengine/godot/issues/27427#issuecomment-664539160), but I was unable to get it to work.

An example of the gap can be seen in this screenshot:

![Screenshot from 2021-02-04 22-46-34](https://user-images.githubusercontent.com/49003204/106966323-01745800-673d-11eb-8233-a28aa739035a.png)

**Minimal reproduction project:**
https://github.com/RobTheFiveNine/godot-collision-polygon-margin-example

**Comment by hoontee** [comment]
Author of the workaround here. Are you sure this code doesn't work?:

```GDScript

func _ready() -> void:
	fix_collision_polygons($"/root/Game/World") # Your nodes to fix here

func fix_collision_polygons(parent: Node) -> void:
	for child in parent.get_children():
		if parent is PhysicsBody and child is CollisionPolygon and not child.disabled:
			var body := (parent as CollisionObject).get_rid()
			for i in PhysicsServer.body_get_shape_count(body):
				var shape := PhysicsServer.body_get_shape(body, i)
				if PhysicsServer.shape_get_type(shape) == PhysicsServer.SHAPE_CONVEX_POLYGON:
					var new_convex_polygon_shape := ConvexPolygonShape.new()
					new_convex_polygon_shape.margin = 0
					new_convex_polygon_shape.set_points(PhysicsServer.shape_get_data(shape))
					var new_collision_shape := CollisionShape.new()
					new_collision_shape.shape = new_convex_polygon_shape
					new_collision_shape.transform = child.transform
					parent.add_child(new_collision_shape)
			child.free()
		else:
			fix_collision_polygons(child)
```

**Comment by RobTheFiveNine** [comment]
> Author of the workaround here. Are you sure this code doesn't work?:
> 
> ```gdscript
> func _ready() -> void:
> 	fix_collision_polygons($"/root/Game/World") # Your nodes to fix here
> 
> func fix_collision_polygons(parent: Node) -> void:
> 	for child in parent.get_children():
> 		if parent is PhysicsBody and child is CollisionPolygon and not child.disabled:
> 			var body := (parent as CollisionObject).get_rid()
> 			for i in PhysicsServer.body_get_shape_count(body):
> 				var shape := PhysicsServer.body_get_shape(body, i)
> 				if PhysicsServer.shape_get_type(shape) == PhysicsServer.SHAPE_CONVEX_POLYGON:
> 					var new_convex_polygon_shape := ConvexPolygonShape.new()
> 					new_convex_polygon_shape.margin = 0
> 					new_convex_polygon_shape.set_points(PhysicsServer.shape_get_data(shape))
> 					var new_collision_shape := CollisionShape.new()
> 					new_collision_shape.shape = new_convex_polygon_shape
> 					new_collision_shape.transform = child.transform
> 					parent.add_child(new_collision_shape)
> 			child.free()
> 		else:
> 			fix_collision_polygons(child)
> ```

I must have made a typo somewhere when trying it last time, because I can't break it this time! Above example and the one from the original post seems to work

**Comment by hoontee** [comment]
Glad we could get it to work. I do a lot of CSG / collision workarounds like this for my projects so let me know if you need any more help.

Technically this issue is a duplicate of #27427 since `CollisionPolygons` decimate into `ConvexPolygonShapes`:

https://github.com/godotengine/godot/blob/cea42faa77d72cdbea8ab7c38936e7c8722900d0/scene/3d/collision_polygon_3d.cpp#L54-L75

**Comment by hoontee** [comment]
#27427 will need to be fixed first since `CollisionPolygons` decimate into `ConvexPolygonShapes`:

https://github.com/godotengine/godot/blob/cea42faa77d72cdbea8ab7c38936e7c8722900d0/scene/3d/collision_polygon_3d.cpp#L54-L75

**Comment by RobTheFiveNine** [comment]
Ahhh OK, this issue can probably be closed then and just added as a note to #27427 then. I didn't realise `CollisionPolygon`s get turned into convex shapes under the hood.

Thanks for the fix too! I ended up scaling everything up in the project I ran into it with, which has actually helped fix some lighting issues too, but I'll definitely keep this solution bookmarked for future use 😃

**Comment by hoontee** [comment]
> Ahhh OK, this issue can probably be closed then and just added as a note to #27427 then.

No, this issue should be kept open as it is separate. It's just that if this gets fixed and a `margin` property is added, it won't have any effect until #27427 is fixed.

**Comment by hoontee** [comment]
I'm sorry, I take it back entirely - adding a `margin` property and setting it to zero will still have an effect if #27427 isn't fixed first...
